### DISCUSSION TITLE
Trouble deploying lambda function with a predefined IAM role

### DISCUSSION THREAD

**Discussion by codeandA** [cargo-lambda/cargo-lambda#342] [Q&A] [answered]
Hello! First and foremost, fantastic project! "cargo-lambda" makes the Rust lambda development experience seamless and definitely one of the nicest on AWS for any language. It really helps unlock Rust’s potential on the platform!

I’m recently having some trouble making my lambdas assume a predefined role already present in my IAM console. I used to be able to use the "--role" flag in the "cargo lambda deploy" command in order to assume the desired role:

```
cargo lambda deploy [bin-name] --role arn:aws:iam::XXXXXXXXXXXXXXXX:role/rust-lambda-role --region us-east-1 --profile [profile-name]
```

Whenever I try that now, I get a "creating execution role" message during deployment, and the lambda generates a default cargo-lambda-role with only the "AWSLambdaBasicExecutionRole" permission attached to it. After that, I have to manually change the role associated with the function to my predefined desired one or possibly change the permissions in the default role.

I tried many variations of the command:

1) Using the "--iam-role" flag mentioned in the official cargo-lambda documentation instead of the "--role" flag mentioned on the cargo lambda deploy --help subcommand
2) Passing only the role identifier name without the full ARN path
3) Enclosing the role's full ARN path in single or double quotes

and combinations of all of them.

The permissions I’m testing the predefined role with are:

"AWSLambdaExecute" and "AmazonS3FullAccess"

Any insight on why the lambda function is not being deployed with the predefined role would be much appreciated.

**Comment by calavera** [comment, marked as answer]
Which version are you using? The function should be attached to the role that you're setting in the flag, it's probably a bug if that doesn't work anymore.

**Reply by codeandA** [reply]
I'm currently using the latest version (0.16), but I started having problems with a previous version (0.14). I'm not sure if maybe it's something new with how AWS receives the information to attach roles to the functions (I'm using version 2.9.0 of aws-cli, btw) or if it's the fact that I'm now explicitly setting the --region and --profile tags, which were default before, and that conflicts with the --role information somehow. 

If you can't reproduce the bug, I can try to give some extra information on how my AWS account is set up if it helps.

**Reply by codeandA** [reply]
Hello @calavera ! Thank you very much for your initial response. I've done some additional testing and realized most configurations I'm setting through tags aren't really being deployed to the environment in AWS (other than the --region one). I've tested environment variables with the tag --env-var, and there are no variables set once the lambda has finished uploading.

I ran a test with the --verbose tag and got the following results:

```
▸▹▹▹▹ loading binary data                                                                                                                                                           DEBUG run:function_deploy_metadata{manifest_path="Cargo.toml" name="rust-lambda"}: searching package metadata name="rust-lambda" target_matches=false
DEBUG run:function_deploy_metadata{manifest_path="Cargo.toml" name="rust-lambda"}: searching package metadata name="rust-lambda" target_matches=false
DEBUG run:function_deploy_metadata{manifest_path="Cargo.toml" name="rust-lambda"}: using deploy configuration from metadata config=None
DEBUG run: loaded function metadata for deployment env=None metadata=None
▹▹▹▹▹ verifying role access, this can take up to 20 seconds                                                                   
```

I feel that those None values are the reason the flags aren't being applied, but I still can't find out what in my development environment is causing this.

At the end, the DEBUG line related to the deployment shows this:

```
config=DeployConfig { memory: None, timeout: Timeout(30), env: {}, env_file: None, tracing: PassThrough, iam_role: None, layers: None, tags: None }
▹▹▹▹▹ deploying function                                                                                                                                                            DEBUG run: uploading zip to Lambda
```

I'll do some further testing, but any ideas are appreciated!

**Reply by calavera** [reply]
can you give me more information about your project:

- What's the exact command that you execute to deploy?
- Do you have any deploy configuration options in the Cargo.toml file? can you post them here with fake values?

**Reply by calavera** [reply]
I think I've traced the problem to a changed we made before 0.14 to not deploy empty environment variables. I'm going to add a few extra tests, but I should have it fixed in a few days.

**Reply by codeandA** [reply]
Oh, I'm glad to hear it! Just as a reference, here is the Cargo.toml file from the lambda I'm using in this test:

```
[package]
name = "rust-lambda"
version = "0.1.0"
edition = "2021"

# Starting in Rust 1.62 you can use `cargo add` to add dependencies 
# to your project.
#
# If you're using an older Rust version,
# download cargo-edit(https://github.com/killercup/cargo-edit#installation) 
# to install the `add` subcommand.
#
# Running `cargo add DEPENDENCY_NAME` will
# add the latest version of a dependency to the list,
# and it will keep the alphabetic ordering for you.

[dependencies]

aws-config = "0.53.0"
aws-sdk-s3 = "0.23.0"
aws-sdk-sqs = "0.23.0"
aws_lambda_events = "0.7.3"
lambda_runtime = "0.7"
serde = "1.0.136"
serde_json = "1.0.91"
tokio = { version = "1", features = ["macros"] }
tracing = { version = "0.1", features = ["log"] }
tracing-subscriber = { version = "0.3", default-features = false, features = ["fmt"] }
```

The mock lambda is very simple; it just writes a message to SQS when a new file arrives at S3, along with some information obtained by processing an S3 Event.

The deployment commands that follow are:

```
cargo lambda build --release
```

```
cargo lambda deploy --env-var QUEUE_URL=https://sqs.us-east-1.amazonaws.com/XXXXXXXXXXXX/Queue.fifo --role arn:aws:iam::XXXXXXXXXXXX:role/rust-lambda-role --region us-east-1 --profile admin --verbose
```

The lambda gets deployed normally; however, looking at it in the AWS console afterwards confirms that the role it assumed was the default one it created and not the one I'm passing to it, and there are no environment variables defined either.

I'm not sure if it matters, but my rustc --version is 1.65.0.

**Reply by calavera** [reply]
I made some updates that landed in `main` last night that I believe fix these problems. I'll probably release a patch version tonight, but if you want to double check, I'll appreciate it. You can install from source by cloning the repository and running `make install-local` in the root of the repo.

**Reply by codeandA** [reply]
It worked perfectly @calavera ! The new version of the CLI was able to set both the correct role and also the environment variables. Thank you very much for the fix and the awesome work with the tool as a whole!

**Reply by calavera** [reply]
awesome! I'm glad I can help.

**Reply by calavera** [reply]
I juse released 0.16.1 with those fixes.
